compiler/rustc_ast_lowering/src/delegation.rs

```diff
@@ -103,29 +134,83 @@ impl<'hir> LoweringContext<'_, 'hir> {
         }
     }
 
-    fn add_inline_attribute_if_needed(&mut self, span: Span) {
-        const PARENT_ID: hir::ItemLocalId = hir::ItemLocalId::ZERO;
-        let create_inline_attr_slice =
-            || [hir::Attribute::Parsed(AttributeKind::Inline(InlineAttr::Hint, span))];
-
-        let new_attributes = match self.attrs.get(&PARENT_ID) {
-            Some(attrs) => {
-                // Check if reuse already specifies any inline attribute, if so, do nothing
-                if attrs
-                    .iter()
-                    .any(|a| matches!(a, hir::Attribute::Parsed(AttributeKind::Inline(..))))
-                {
-                    return;
+    fn add_attributes_if_needed(&mut self, span: Span, sig_id: DefId) {
+        let mut candidate_additions = ATTRIBUTES_ADDITIONS
+            .iter()
+            .map(|info| AttributeAdditionState { info, already_exists: false })
+            .collect::<Vec<_>>();
+
+        self.mark_existing_attributes(&mut candidate_additions);
+        let new_attributes = self.create_new_attributes(&candidate_additions, span, sig_id);
+
+        if new_attributes.is_empty() {
+            return;
+        }
+
+        let new_arena_allocated_attributes = match self.attrs.get(&PARENT_ID) {
+            Some(existing_attrs) => self.arena.alloc_from_iter(
+                existing_attrs.iter().map(|a| a.clone()).chain(new_attributes.into_iter()),
+            ),
+            None => self.arena.alloc_from_iter(new_attributes.into_iter()),
+        };
+
+        self.attrs.insert(PARENT_ID, new_arena_allocated_attributes);
+    }
+
+    fn mark_existing_attributes(&mut self, candidate_additions: &mut Vec<AttributeAdditionState>) {
+        let Some(existing_attrs) = self.attrs.get(&PARENT_ID) else { return };
+
+        for attr in existing_attrs.iter() {
+            for addition in candidate_additions.iter_mut() {
```

Comment:
```suggestion
            for addition in &mut candidate_additions {
```
This form is typically preferred.
(Here and in other places.)